astropy/time/core.py

```diff
@@ -984,24 +991,35 @@ def to_value(self, format, subfmt='*'):
             else:
                 tm = self.replicate(format=format)
 
-            # Custom TimeFormat subclasses created in astropy versions <= 4.0
-            # may not be able to handle being passes on a out_subfmt.
-            # But those do supposedly deal with `self.out_subfmt` internally,
-            # so if subfmt is the same, we do not pass it on.
+            # Some TimeFormat subclasses may not be able to handle being passes
+            # on a out_subfmt. This includes some core classes like
+            # TimeBesselianEpochString that do not have any allowed subfmts. But
+            # those do deal with `self.out_subfmt` internally, so if subfmt is
+            # the same, we do not pass it on.
             kwargs = {}
             if subfmt is not None and subfmt != tm.out_subfmt:
                 kwargs['out_subfmt'] = subfmt
             try:
-                value = tm._shaped_like_input(tm._time.to_value(
-                    parent=tm, **kwargs))
-            except ValueError as exc:
-                if 'subformats match' in str(exc):
-                    raise ValueError('subformat must match one of {}'
-                                     .format([option[0] for option in
-                                              tm._time.subfmts])) from None
+                value = tm._time.to_value(parent=tm, **kwargs)
+            except TypeError as exc:
+                # Try validating subfmt, e.g. for formats like 'jyear_str' that
+                # do not implement out_subfmt in to_value() (because there are
+                # no allowed subformats).  If subfmt is not valid this gives the
+                # same exception as would have occurred if the call to
+                # `to_value()` had succeeded.
+                tm._time._select_subfmts(subfmt)
+
+                # Subfmt was valid, so fall back to the original exception to see
+                # if it was lack of support for out_subfmt as a call arg.
+                if "unexpected keyword argument 'out_subfmt'" in str(exc):
```

Comment:
Perhaps this is a separate issue to the one addressed in this PR, but shouldn't we do more than raise an exception here? i.e shouldn't the following actually work, rather than raise an exception?

```python
t = Time(58837.72643888889, format='mjd')
t.to_value('iso', subfmt='date_hm') 
```

Reply:
Good catch, that is definitely a case that should work!  That is failing in 4.0 as well.  Looks like the existing test coverage for `to_value` needs some work since this version is passing tests now.  On it...